crates/core/src/network/connection_manager.rs

```diff
@@ -708,54 +719,113 @@ where
-        // Draining of incoming queue
-        while let Some(Some(Ok(msg))) = incoming.next().now_or_never() {
-            // ignore malformed messages
-            let Some(header) = msg.header else {
-                continue;
-            };
-            if let Some(body) = msg.body {
-                // we ignore non-deserializable messages (serde errors, or control signals in drain)
-                if let Ok(msg) = body.try_as_binary_body(protocol_version) {
-                    drain_counter += 1;
-                    let parent_context = header.span_context.as_ref().map(|span_ctx| {
-                        global::get_text_map_propagator(|propagator| propagator.extract(span_ctx))
-                    });
-
-                    if let Err(e) = router
-                        .call(
-                            Incoming::from_parts(
-                                msg,
-                                // This is a dying connection, don't pass it down.
-                                WeakConnection::new_closed(peer_node_id),
-                                header.msg_id,
-                                header.in_response_to,
-                                PeerMetadataVersion::from(header),
-                            )
-                            .with_parent_context(parent_context),
-                            protocol_version,
-                        )
-                        .await
-                    {
-                        debug!(
-                            "Error processing message while draining connection: {:?}",
-                            e
+        loop {
+            tokio::select! {
+                _ = &mut drain_timeout => {
+                    // drain timed out.
+                    debug!("Drain timed out, closing connection");
+                    drop_egress.take();
+                    drop(incoming);
+                    break;
+                },
+                msg = incoming.next() => {
+                    let Some(Ok(msg)) = msg else {
+                        break;
+                    };
+                    // Draining of incoming queue
+                    if let Some(Body::ConnectionControl(ctrl_msg)) = &msg.body {
+                        trace!(
+                            "Received signal from {} while draining: {}",
+                            peer_node_id,
+                            ctrl_msg.signal()
                         );
+                        // No more requests/unary messages will arrive. Reactor can be closed if we don't
+                        // have waiting responses
+                        match ctrl_msg.signal() {
+                            Signal::RequestStreamDrained => {
+                                // No more requests coming, but rpc responses might still arrive.
+                                // We don't need tx anymore. We'll not create future responder
+                                // tasks.
+                                tx.take();
```

Comment:
Could we theoretically drop `tx` at the start of this loop? It looks as if we don't do anything with it in the loop body nor afterwards.

Reply:
It'll be used in future work.